Description:
 Fix Denial of Service with IPv6 Router Advertisements [SA-15:09] (CVE-2015-2923)
Origin: vendor, https://security.FreeBSD.org/patches/SA-15:09/ipv6.patch
Bug: https://www.freebsd.org/security/advisories/FreeBSD-SA-15:09.ipv6.asc
Bug-Debian: https://bugs.debian.org/XXXXXX
Applied-Upstream: https://svnweb.freebsd.org/base?view=revision&revision=281232

--- a/sys/netinet6/nd6_rtr.c
+++ b/sys/netinet6/nd6_rtr.c
@@ -296,8 +296,16 @@
 	}
 	if (nd_ra->nd_ra_retransmit)
 		ndi->retrans = ntohl(nd_ra->nd_ra_retransmit);
-	if (nd_ra->nd_ra_curhoplimit)
-		ndi->chlim = nd_ra->nd_ra_curhoplimit;
+	if (nd_ra->nd_ra_curhoplimit) {
+		if (ndi->chlim < nd_ra->nd_ra_curhoplimit)
+			ndi->chlim = nd_ra->nd_ra_curhoplimit;
+		else if (ndi->chlim != nd_ra->nd_ra_curhoplimit) {
+			log(LOG_ERR, "RA with a lower CurHopLimit sent from "
+			    "%s on %s (current = %d, received = %d). "
+			    "Ignored.\n", ip6_sprintf(ip6bufs, &ip6->ip6_src),
+			    if_name(ifp), ndi->chlim, nd_ra->nd_ra_curhoplimit);
+		}
+	}
 	dr = defrtrlist_update(&dr0);
     }
 
